docs: Vue tracing integration (#1757)
getsentry/sentry-docs: src/collections/_documentation/performance/distributed-tracing.md

  })
});
```

### Vue.js

> Vue Tracing Integration allows us to track rendering performance during initial application load.

Reviewer: ```suggestion
The Vue Tracing Integration allows us to track rendering performance during an initial application load.
```

> It is achieved by injecting our handler inside Vue's `beforeCreate` mixin, which gives us an access to component during it is all lifecycle stages.

Reviewer: ```suggestion
It is achieved by injecting our handler inside Vue's `beforeCreate` mixin, which gives us access to a Vue component during all of it's lifecycle stages.
```

> When we encounter component named `root`, which is top-level Vue instance (as in `new Vue({})`), we use our APM Tracing integration,

Reviewer: ```suggestion
When we encounter component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM Tracing integration,
```

Think we call it `AM ` instead of `APM` now.

> When we encounter component named `root`, which is a top-level Vue instance (as in `new Vue({})`), we use our AM Tracing integration,

Author: Correct. I'll make sure to only apply changes that are using correct meaning and leave comment otherwise.

> and create a new activity named `Vue Application Render`. Once it is created, it will wait until all it is child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.

Reviewer: ```suggestion
and create a new activity named `Vue Application Render`. Once it is created, it will wait until all of it's child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.
```

> and create a new activity named `Vue Application Render`. Once it is created, it will wait until all of it's child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.

Author: "it is created" = "the activity is created" — and the activity is the thing that will wait for the rest of components.

> and create a new activity named `Vue Application Render`. Once it is created, it will wait until all of it's child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.

Author: ```suggestion
and create a new activity named `Vue Application Render`. Once the activity has been created, it will wait until all of its child components render, and there aren't new rendering events triggered within the configured `timeout`, before marking the activity as completed.
```

> and create a new activity named `Vue Application Render`. Once it is created, it will wait until all of it's child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.

Author: Added my own change to this paragraph, please verify :)

> and create a new activity named `Vue Application Render`. Once it is created, it will wait until all of it's child components are rendered, and there are no new rendering events triggered within configured `timeout`, before marking the activity as completed.

Reviewer 2: Looks good to me! 👍

> This will give you a very high-level information about rendering timing of the app. However, we can provide more fine-grained details about what actually happened during mentioned activity.

Reviewer: ```suggestion
This will give you very high-level information about the rendering performance of the Vue instance, however the integration can also provide more fine-grained details about what actually happened during a specific activity.
```

> This will give you very high-level information about the rendering performance of the Vue instance, however the integration can also provide more fine-grained details about what actually happened during a specific activity.

Reviewer 2: ```suggestion
`Vue Application Render` will give you very high-level information about the rendering performance of the Vue instance. However, the integration can also provide more fine-grained details about what actually happened during a specific activity.
```

Author: "This" = the setup described above, a whole functionality.

Author: `Vue Application Render` is just a name of the activity (a conceptual object that keeps track of all the necessary data)

Reviewer 2: Nice. Thanks for the clarification. 😄

Reviewer 2: I checked the updated paragraph on line 627 and it looks good 👍

> This will give you very high-level information about the rendering performance of the Vue instance, however the integration can also provide more fine-grained details about what actually happened during a specific activity.

Author: ```suggestion
The presented instrumentation will give you very high-level information about the rendering performance of the Vue instance. However, the integration can also provide more fine-grained details about what actually happened during a specific activity.
```

> In order to do that, you need to specify which components you want to track and what hooks you want to listen to (list of all available hooks can be found here https://vuejs.org/v2/api/#Options-Lifecycle-Hooks). If you really want to, you can turn on tracking for all the components, however it can be rather noisy if your app consists of hundreds of components, thus we encourage being more specific. When hooks are not provided, we will track component's `mount` and `update` hooks.

Reviewer 2: ```suggestion
To do that, you need to specify which components you want to track and what hooks you want to listen to (you can find a list of all available hooks can [here](https://vuejs.org/v2/api/#Options-Lifecycle-Hooks)). You can also turn on tracking for all the components. However, it can be rather noisy if your app consists of hundreds of components. We encourage being more specific. If you don't provide hooks, Sentry will track a component's `mount` and `update` hooks.
```

> Note that we do not use `before` and `-ed` pairs for hooks, and you should provide a simple verb instead, eg. `update`, not `beforeUpdate` and `updated`.

Reviewer 2: ```suggestion
Note that we don't use `before` and `-ed` pairs for hooks, and you should provide a simple verb instead. For example, `update` is correct. `beforeUpdate` and `updated` are incorrect.
```

> To setup Vue Tracing Integration, first you need to configure APM Tracing integration itself. For details on how to do this, see [JavaScript]({%- link _documentation/performance/distributed-tracing.md -%}#javascript) section above.

Reviewer: ```suggestion
To setup the Vue Tracing Integration, you will first need to configure the AM Tracing integration itself. For details on how to do this, see [JavaScript]({%- link _documentation/performance/distributed-tracing.md -%}#javascript) section above.
```

> To setup the Vue Tracing Integration, you will first need to configure the AM Tracing integration itself. For details on how to do this, see [JavaScript]({%- link _documentation/performance/distributed-tracing.md -%}#javascript) section above.

Reviewer 2: ```suggestion
To set up the Vue Tracing Integration, you will first need to configure the AM Tracing integration itself. For details on how to do this, see the [JavaScript]({%- link _documentation/performance/distributed-tracing.md -%}#javascript) section above.
```

> Once it is done, you can move to configuring Vue integration itself.

Reviewer: ```suggestion
Once the Tracing integration has been configured, you can move on to configuring the Vue integration itself.
```

> Once the Tracing integration has been configured, you can move on to configuring the Vue integration itself.

Reviewer 2: ```suggestion
Once you've configured the Tracing integration, move on to configuring the Vue integration itself.
```

> Our new tracing capabilites are baked-in inside original Vue error handler integrations, so there is no need to add any new packages. You only need to provide an appropriate configuration.

Reviewer 2: ```suggestion
Sentry's new tracing capabilities are built into the original Vue error handler integrations, so there is no need to add any new packages. You only need to provide an appropriate configuration.
```

> The most basic configuration for the tracing, which would track only top-level component looks like this:

Reviewer: ```suggestion
The most basic configuration for tracing your Vue app, which would track only the top-level component, looks like this:
```

```js
import * as Sentry from "@sentry/browser";
import { Vue as VueIntegration } from "@sentry/integrations";
import { Integrations } from "@sentry/apm";
import Vue from "vue";

Sentry.init({
  // ...
  integrations: [
    new Integrations.Tracing(),
    new VueIntegration({
      Vue,
      tracing: true
    })
  ],
  tracesSampleRate: 1
});
```

> If we want to track child components, and see more details about the rendering process, we can ask the integration to either track them all:

Reviewer 2: ```suggestion
If you want to track child components, and see more details about the rendering process, configure the integration to track them all:
```

```js
new VueIntegration({
  Vue,
  tracing: true,
  tracingOptions: {
    trackComponents: true
  }
})
```

> or be more granular with our choices:

Reviewer 2: ```suggestion
Or, you can choose more granularity:
```

```js
new VueIntegration({
  Vue,
  tracing: true,
  tracingOptions: {
    trackComponents: ["App", "RwvHeader", "RwvFooter", "RwvArticleList", "Pagination"]
  }
})
```

> If we also want to know if some components are for example removed during initial page load, we can add `destroy` hook, to our defaults:

Reviewer 2: ```suggestion
If you want to know if some components are, for example, removed during the initial page load, add a `destroy` hook to the default:
```

```js
new VueIntegration({
  Vue,
  tracing: true,
  tracingOptions: {
    trackComponents: ["App", "RwvHeader", "RwvFooter", "RwvArticleList", "Pagination"],
    hooks: ['mount', 'update', 'destroy']
  }
})
```

> The last thing that we can specify, is how long top-level activity should wait for the last component to perform it is rendering.

Reviewer 2: ```suggestion
You can specify how long a top-level activity should wait for the last component to render.
```

> Every new rendering cycle is debouncing the timeout, and it starts counting from the beginning. Once the timeout is reached, tracking is completed and all the information is sent to Sentry.

Reviewer 2: ```suggestion
Every new rendering cycle is debouncing the timeout, and it starts counting from the beginning. Once the timeout is reached, tracking is completed, and all the information is sent to Sentry.
```

```js
new VueIntegration({
  Vue,
  tracing: true,
  tracingOptions: {
    trackComponents: true,

>     timeout: 4000

Reviewer: How does this interact with the Tracing integrations `idleTimeout` option?

Author: This timeout pops the activity itself, so `idleTimeout` won't fire up because there'll still be something pending.

  }
})
```

#### Configuration

```js
/**
 * When set to `true`, enables tracking of components lifecycle performance.
 * Default: false
 */
tracing: boolean;
tracingOptions: {
  /**
   * Decides whether to track components by hooking into its lifecycle methods.
   * Can be either set to `boolean` to enable/disable tracking for all of them.
   * Or to an array of specific component names (case-sensitive).
   * Default: false
   */
  trackComponents: boolean | string[];
  /**

>    * How long to wait (in ms) until the tracked root activity is marked as finished and sent of to Sentry

Reviewer 2: ```suggestion
   * How long to wait (in ms) until the tracked root activity is marked as finished and sent to Sentry
```

   * Default: 2000
   */
  timeout: number;
  /**
   * List of hooks to keep track of during component lifecycle.
   * Available hooks: 'activate' | 'create' | 'destroy' | 'mount' | 'update'
   * Based on https://vuejs.org/v2/api/#Options-Lifecycle-Hooks
   */

>   hooks: string[];

Reviewer: Should we explicitly write the pairs somewhere?

```js
const HOOKS: { [key in Operation]: Hook[] } = {
  activate: ['activated', 'deactivated'],
  create: ['beforeCreate', 'created'],
  destroy: ['beforeDestroy', 'destroyed'],
  mount: ['beforeMount', 'mounted'],
  update: ['beforeUpdate', 'updated'],
};
```

Author: I'm not sure if this won't be unnecessary noise. It's rather self-explanatory.
